components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt

```diff
@@ -7,35 +7,116 @@ package mozilla.components.feature.downloads
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
 import mozilla.components.browser.state.action.BrowserAction
 import mozilla.components.browser.state.action.DownloadAction
 import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.content.DownloadState
+import mozilla.components.browser.state.state.content.DownloadState.Status.CANCELLED
+import mozilla.components.browser.state.state.content.DownloadState.Status.COMPLETED
 import mozilla.components.lib.state.Middleware
 import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+import mozilla.components.support.base.log.logger.Logger
+import kotlin.coroutines.CoroutineContext
 
 /**
  * [Middleware] implementation for managing downloads via the provided download service. Its
  * purpose is to react to global download state changes (e.g. of [BrowserState.downloads])
  * and notify the download service, as needed.
  */
+@Suppress("ComplexMethod")
 class DownloadMiddleware(
     private val applicationContext: Context,
-    private val downloadServiceClass: Class<*>
+    private val downloadServiceClass: Class<*>,
+    coroutineContext: CoroutineContext = Dispatchers.IO,
+    @VisibleForTesting
+    internal val downloadStorage: DownloadStorage = DownloadStorage(applicationContext)
 ) : Middleware<BrowserState, BrowserAction> {
+    private val logger = Logger("DownloadMiddleware")
 
+    private var scope = CoroutineScope(coroutineContext)
+
+    @InternalCoroutinesApi
     override fun invoke(
         context: MiddlewareContext<BrowserState, BrowserAction>,
         next: (BrowserAction) -> Unit,
         action: BrowserAction
     ) {
-        next(action)
-
         when (action) {
             is DownloadAction.AddDownloadAction -> {
-                val intent = Intent(applicationContext, downloadServiceClass)
-                intent.putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, action.download.id)
-                applicationContext.startService(intent)
+                next(action)
+                scope.launch {
+                    downloadStorage.add(action.download)
+                    logger.debug("Added download ${action.download.fileName} to the storage")
+                }
+                sendDownloadIntent(action.download)
             }
+
+            is DownloadAction.RemoveDownloadAction -> removeDownload(action.downloadId, context.store)
+
+            is DownloadAction.RemoveAllDownloadsAction -> removeDownloads()
+
+            is DownloadAction.UpdateDownloadAction -> updateDownload(action.download, context.store)
+
+            is DownloadAction.RestoreDownloadsStateAction -> restoreDownloads(context.store)
+
+            is DownloadAction.RestoreDownloadStateAction -> {
+                next(action)
+                sendDownloadIntent(action.download)
+            }
+        }
+        if (action !is DownloadAction.AddDownloadAction) {
```

Comment:
That's a bug here now. `RestoreDownloadStateAction` would cause next to be called twice. I would handle this differently and have a `preProcess` function that returns whether nor not `next` should be called e.g. whether or not the original action should hit the store.

Here's an example:  https://github.com/mozilla-mobile/android-components/blob/master/components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewMiddleware.kt#L55

Reply:
Thanks for catching this 👍

Reply:
What do you think about 
```kotlin
        if (requireNext(context, action)) {
            next(action)
            postNextProcess(action)
        }
```

Reply:
I find that a bit confusing, to be honest :).

Reply:
So, I would check actions individually or do what we did in the `ReaderViewMiddleware` so we follow the same pattern.